### ISSUE TITLE
The header icons are not visible when choosing a light background

### ISSUE THREAD

**Issue by szaimen** [nextcloud/server#35482]
How can we work around the following issue?

The header icons are not visible when choosing a light background image (same for admin chosen background image):
![image](https://user-images.githubusercontent.com/42591237/204506475-0edcbc2d-db96-473a-9196-b2dde9e7036e.png)

What about adding a switch that allows to invert the header icons? (I fear Ii don't see any other good solution to automate this)

cc @nimishavijay @jancborchardt

**Comment by jancborchardt** [comment]
I mean, they are also not properly visible if you pick a checkerboard background or something like that.

If you pick a custom background, the automatic color extraction by @skjnldsv will mostly take care of it (and set e.g. white as background color, so the icons would then be dark).

However if the majority of the image is not white, it can still be fixed by changing the color. No additional setting needed I'd say.

**Comment by szaimen** [comment]
> If you pick a custom background, the automatic color extraction by @skjnldsv will mostly take care of it (and set e.g. white as background color, so the icons would then be dark).

I fear this is not how it works in my testing unfortunately... See the screenshot I've sent.

**Comment by skjnldsv** [comment]
> If you pick a custom background, the automatic color extraction by @skjnldsv will mostly take care of it (and set e.g. white as background color, so the icons would then be dark).

No, this will not work reliably.
Have a look at [an image like this](https://user-images.githubusercontent.com/14975046/204592612-3a9fc5b3-c2c7-487b-ba68-b50001c9b55d.jpg), bright top-left header, so icons are not visible. But overall luminance is nowhere bright :)
Such edge cases are impossible to automate imho

![image](https://user-images.githubusercontent.com/14975046/204592512-a9e26780-f6cd-439a-b916-520d3c0eba59.png)

**Comment by skjnldsv** [comment]
> However if the majority of the image is not white, it can still be fixed by changing the color. No additional setting needed I'd say.

The invert of the top header is not based on the primary colour.

**Comment by szaimen** [comment]
> What about adding a switch that allows to invert the header icons? (I fear Ii don't see any other good solution to automate this)

What about this idea?

**Comment by skjnldsv** [comment]
> > What about adding a switch that allows to invert the header icons? (I fear Ii don't see any other good solution to automate this)
> 
> What about this idea?

For me the only one that would work.
Automating this is not easily doable unfortunately

**Comment by florianwgnr** [comment]
Duplictates #35300

**Comment by szaimen** [comment]
> > > What about adding a switch that allows to invert the header icons? (I fear Ii don't see any other good solution to automate this)
> > 
> > 
> > What about this idea?
> 
> For me the only one that would work. Automating this is not easily (if at all) doable unfortunately

@jancborchardt would this be fine for you as well?

**Comment by florianwgnr** [comment]
It seems as if this issue is already fixed within the nextcloud instant trial: 
Fresh installation (manually):
![grafik](https://user-images.githubusercontent.com/39341393/205487684-51935e7b-6651-4097-8637-ace0a4614652.png)

Nextcloud Instant Trial:
![grafik](https://user-images.githubusercontent.com/39341393/205487694-76421588-ffdc-4986-82e7-0e8a486fb7aa.png)

(Firefox 107)

**Comment by susnux** [comment]
I do not think inverting the color does work here, as soon as you allow to set a background image you can not assume one color would work on the whole width of the menu bar.

I think the only solution working for every image is to not render the icon directly on the background but use a layer between with a fixed background like the dashboard panels.

This is also true for all other places where we render text directly on the background like the footer text on the login page, the site title on the header of public template or the "hello, NAME" on the dashboard.

So maybe we could add a option to enable a background on the header. I have created a mockup, it could look like this:
|bright color mode|
|:---:|
|![header-bright-mode](https://user-images.githubusercontent.com/1855448/235905006-238ab7d7-0bd4-4cc5-a8b1-c1ffd3325546.png)|
|**dark color mode**|
|![header-darkmode](https://user-images.githubusercontent.com/1855448/235905012-2ebe7ed0-df07-4232-800e-d2b951490a82.png)|

**Comment by susnux** [comment]
After slept a night to think about it: While this fixes the dashboard, for every other app I am not sure if this is design-wise good :confused: 

files app | files app dark | files app with margin
---|---|---
![rounded background on the headerbar](https://user-images.githubusercontent.com/1855448/236195858-2423462b-88a1-479a-a08f-03a96643747e.png)|![rounded background on the header bar using dark color mode](https://user-images.githubusercontent.com/1855448/236195884-c4b87d10-bb69-42c3-ab99-6dc81fdb1f31.png)|![giving the rounded background on the headerbar some margin on the top and bottom](https://user-images.githubusercontent.com/1855448/236195900-efcb55b0-bf07-452e-8da9-1a57b04c9c64.png)

**Comment by nimishavijay** [comment]
Yes, these are edge cases and unfortunately it would be difficult to make sure all cases are taken care of perfectly through automation. A switch for icons sounds the best to me as well, it can be a toggle with "Invert icon colors in the header" and it is off by default every time a new background is chosen.

A background for all the other items where there may be visibility issues would look like this:
![image](https://user-images.githubusercontent.com/52440189/236207664-5f35c78b-86c4-42ea-8795-ecc969c4982e.png)

<img src="https://user-images.githubusercontent.com/52440189/236209001-310b9664-5e1f-42ee-a6e3-7296fd1f5526.png" width=50% height=50%>

I am not sure if this is a design we want. Perhaps if the background is causing so many issues it's simpler to just change the background :) cc @jancborchardt

**Comment by susnux** [comment]
I agree, so just adding the background on the greeting (dashboard) + login page would allow much more images so that adding an invert-icons switch should work :)

Especially on the login page we need an background, as even for the default image it breaks (currently only works for some users, see issue https://github.com/nextcloud/server/issues/36950 ).

**Comment by jancborchardt** [comment]
I think on the login page yep that’s fine and needed, as you say @susnux. The "Welcome" text starts looking a bit strange and ugly with the container around. We could make the text a tiny bit bigger as well, which can also fix contrast issues.

And also tbh, custom backgrounds are not covered by contrast requirements. As said if there’s a checkerboard background, or a screenshot or similar, then that’s just not something you should choose. That’s why I would say we shouldn’t do anything like a container for the big "Welcome" text or the apps header.
